caikit/core/data_model/streams/multipart_decoder.py

```diff
@@ -0,0 +1,120 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""A decoder for files that contain multi-part form data of files.
+
+e.g. files of the form
+```
+--my_boundary
+Content-Disposition: form-data; name=""; filename="foo.json"
+Content-Type: application/json
+
+[
+    {
+        "foo": "bar"
+    }
+]
+
+--my_boundary--
+```
+
+This is not meant to be an implementation detail of an HTTP handler, this is only meant to be a
+general-purpose file decoder for cases where:
+- A multipart form-data request has been serialized to a file
+- That form data itself contained one or more files (NB: _not_ form fields)
+- We wish to get a stream of those inner file contents back
+
+Currently implemented using the `werkzeug` multipart parser, for ease of implementation.
+
+A known drawback is that this reads the full file before streaming back the parsed file contents.
+Reading multi-GB files could be slow, as this will cause serialization back to disk of the
+files contained within the multipart request.
+"""
+# Standard
+from typing import Iterator
+import dataclasses
+import os
+import typing
+
+# Third Party
+from werkzeug import formparser
+from werkzeug.datastructures import FileStorage
+
+# First Party
+from alog import alog
+
+# Local
+from caikit.core.exceptions import error_handler
+
+log = alog.use_channel("MULTIPART_DECODER")
+error = error_handler.get(log)
+
+
+@dataclasses.dataclass
+class Part:
+    content_type: str
+    filename: str
+    fp: typing.IO[bytes]
+
+
+def is_multipart_file(file) -> bool:
+    """Returns true if the file appears to contain a multi-part form data request"""
+    with open(file, "r", encoding="utf-8") as fp:
+        # Read a small bit of the file
+        head: str = fp.read(50)
+
+    # The beginning of the file content should start with "--"
+    return head.lstrip().startswith("--")
```

Comment:
Yuck! What about

```yaml
---
foo: 1
bar: 2
---
foo: 3
bar: 4
```

We currently don't support multi-yaml, but I wouldn't be _shocked_ if that comes down the pike

Reply:
Reference: https://github.ibm.com/watson-vision/ServerConfig/blob/d97cab43378c9031a4cce6d86ce7c9016ff2fa7a/src/trans_inst_image_base.cpp#L494

Reply:
Yeah....
Started an internal convo with some users that are supplying files without the boundary header. We should probably also check for the initial content-type header though as well